src/vmm/src/builder.rs

```diff
@@ -763,39 +767,568 @@ pub mod tests {
+        fn as_raw_fd(&self) -> RawFd {
+            self.0.as_raw_fd()
+        }
     }
+    impl devices::legacy::ReadableFd for SerialInput {}
 
-    #[test]
-    fn test_setup_serial_device() {
-        // Wrapper over TempFile that implements `ReadableFd`.
-        struct SerialInput(File);
-        impl io::Read for SerialInput {
-            fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-                self.0.read(buf)
-            }
+    fn default_kernel_file() -> File {
+        let path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        let kernel_path = path.parent().unwrap().join("kernel/src/loader");
+
+        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+        {
+            File::open(kernel_path.join("test_elf.bin")).unwrap()
         }
-        impl AsRawFd for SerialInput {
-            fn as_raw_fd(&self) -> RawFd {
-                self.0.as_raw_fd()
-            }
+
+        #[cfg(target_arch = "aarch64")]
+        {
+            File::open(kernel_path.join("test_pe.bin")).unwrap()
         }
-        impl devices::legacy::ReadableFd for SerialInput {};
+    }
 
+    fn default_mmio_device_manager() -> MMIODeviceManager {
+        MMIODeviceManager::new(
+            &mut (arch::MMIO_MEM_START as u64),
+            (arch::IRQ_BASE, arch::IRQ_MAX),
+        )
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    fn default_portio_device_manager(event_manager: &mut EventManager) -> PortIODeviceManager {
         let read_tempfile = TempFile::new("/tmp/serial_read_").unwrap();
         let read_file = File::open(read_tempfile.as_path()).unwrap();
         let read_handle = SerialInput(read_file);
+
+        let serial =
+            setup_serial_device(event_manager, Box::new(read_handle), Box::new(io::stdout()))
+                .unwrap();
+        let i8042_reset_evfd = EventFd::new(libc::EFD_NONBLOCK).unwrap();
+        PortIODeviceManager::new(serial, i8042_reset_evfd).unwrap()
+    }
+
+    fn default_kernel_cmdline() -> Cmdline {
+        let mut kernel_cmdline = kernel::cmdline::Cmdline::new(4096);
+        kernel_cmdline.insert_str(DEFAULT_KERNEL_CMDLINE).unwrap();
+        kernel_cmdline
+    }
+
+    fn default_boot_config() -> BootConfig {
+        BootConfig {
+            cmdline: default_kernel_cmdline(),
+            kernel_file: default_kernel_file(),
+        }
+    }
+
+    fn default_vmm(event_manager: &mut EventManager) -> Vmm {
+        let guest_memory = create_guest_memory(128).unwrap();
+        let kernel_cmdline = default_kernel_cmdline();
+
+        let exit_evt = EventFd::new(libc::EFD_NONBLOCK)
+            .map_err(Error::EventFd)
+            .map_err(StartMicrovmError::Internal)
+            .unwrap();
+
+        let vm = setup_kvm_vm(&guest_memory).unwrap();
+        let mmio_device_manager = default_mmio_device_manager();
+        #[cfg(target_arch = "x86_64")]
+        let pio_device_manager = default_portio_device_manager(event_manager);
```

Comment:
We don't need a functioning `serial` for tests just like we don't need a functioning `i8042`.

This can be replaced by:
```rust
PortIODeviceManager::new(
    Arc::new(Mutex::new(Serial::new_sink(EventFd::new(libc::EFD_NONBLOCK).unwrap()))),
    EventFd::new(libc::EFD_NONBLOCK).unwrap()
).unwrap()
```